### DISCUSSION TITLE
Debian-slim as Alpine alternative concerning disk footprint

### DISCUSSION THREAD

**Discussion by pr-ju** [lima-vm/lima#1582] [Ideas]
Since ssd space is often an issue on macs but Alpine uses a different c library, which can result in poor perforce when running a python app ([https://pythonspeed.com/articles/alpine-docker-python/](https://pythonspeed.com/articles/alpine-docker-python/)), I would recommend supporting debian-slim as alternative. With debian-slim the footprint is ca. 82 MiB, which is larger than Alpine but still okay. Maybe Ubuntu minimal is another option here.

**Comment by afbjorklund** [comment]
You can use _any_ image on the VM, the libc is included in the container. Performance issues with it (the VM), would be in the kernel (that is shared) - and of course in the hypervisor. The libc on the VM only affects the apps on the VM, in this case that is **not** python.

**Reply by pr-ju** [reply]
I just thought about the size of the vm disk, since all other images downloaded by the templates are quite big (> 200 or 300 GiB)

**Reply by afbjorklund** [reply]
I think you mean MiB, they are not _that_ big :-) But yes, 300M for the "small" ones and 600M for the "big" ones.

When compared with something like CoreOS (1G) or OpenShift (10G), those sizes are not so big - even on the SSD

**Reply by afbjorklund** [reply]
Actually the debian images were slightly smaller (half), so if you are concerned about disk space it might be worth it.

```
#	/home/anders/.cache/lima/download/by-url-sha256/65a62e4461082eeb92c0623f4db7ba6b27ea625b5e91ff7ddbdca2d721302777
728M	https://cloud-images.ubuntu.com/releases/22.10/release-20230413/ubuntu-22.10-server-cloudimg-amd64.img
#	/home/anders/.cache/lima/download/by-url-sha256/37f5fc187ab151950c3e7f30e2231bedf2f8af23b24933aec9f3a2a2020fa4ba
319M	https://cloud.debian.org/images/cloud/bullseye/20230124-1270/debian-11-generic-amd64-20230124-1270.qcow2
```

Made an issue to include the size with the checksums, to make it easier to know the size "ahead of time" - to compare

* #1586

**Reply by afbjorklund** [reply]
Note that the above sizes are _without_ the container runtime, so that needs to be added too for a fair comparison.

```
#	/home/anders/.cache/lima/download/by-url-sha256/e22c9d3619acd34258d3d43e2b685bf6d4b421aa5c83f4241d96b1ee6b6fe141
233M	https://github.com/containerd/nerdctl/releases/download/v1.3.0/nerdctl-full-1.3.0-linux-amd64.tar.gz
#	/home/anders/.cache/lima/download/by-url-sha256/28137234cfb6f64a587565fb9676c1ddf73912e99520e5026ef9bca0672b7ce0
62M	https://github.com/lima-vm/alpine-lima/releases/download/v0.2.27/alpine-lima-std-3.17.0-x86_64.iso
```

It would be possible to create a custom debian-lima image, that included all the software and used dynamic linking.

Most likely it would end up being somewhere in-between 50-150M (alpine) and 300-600M (debian), in download ?

**Reply by afbjorklund** [reply]
Currently there is no debian packaging for nerdctl and for buildkit (as far as I know), so that could be a start.

* https://github.com/containerd/nerdctl/issues/1986

A workaround could be to just package the nerdctl-full binaries  as debian packages, and install with `apt`.

```
8.8M	/tmp/lima/buildg_0.4.1_amd64.deb
38M	/tmp/lima/buildkit_0.11.6_amd64.deb
3.9M	/tmp/lima/bypass4netns_0.3.0_amd64.deb
33M	/tmp/lima/cni-plugins_1.3.0_amd64.deb
39M	/tmp/lima/containerd_1.7.1_amd64.deb
4.1M	/tmp/lima/containerd-fuse-overlayfs_1.0.6_amd64.deb
744K	/tmp/lima/fuse-overlayfs_1.12.0_amd64.deb
9.2M	/tmp/lima/nerdctl_1.4.0_amd64.deb
12K	/tmp/lima/nerdctl-rootless-extras_1.4.0_amd64.deb
12M	/tmp/lima/rootlesskit_1.1.0_amd64.deb
3.9M	/tmp/lima/runc_1.1.7_amd64.deb
1020K	/tmp/lima/slirp4netns_1.2.0_amd64.deb
376K	/tmp/lima/tini_0.19.0_amd64.deb
152M	total
```

**Comment by afbjorklund** [comment]
> Maybe Ubuntu minimal is another option here.

Maintaining a special optimized Ubuntu cloud image for Lima _could_ be a good idea, but it also needs to be maintained properly.

It would be cool if any OS vendor did a special "spin" for it, but that can be done with a custom yaml? Or we could make our own OS.

But the [Debian cloud images](https://cloud.debian.org/images/cloud/) (qcow2) are smaller

**Reply by afbjorklund** [reply]
I made a custom OS for Lima called [Bean](https://github.com/afbjorklund/bean), and it would open up for have multiple flavors of beans - not just green...

The default variant was optimized on size, so it was also using [musl](https://musl.libc.org/) - just like Alpine. But we could build one with glibc.

The basic operating system doesn't have any package manager, the user is supposed to run containers on it (or `toolbox`)

Using `cloud-init` is major source of bloat and hassle, so it should be replaced with one of the alternative implementations.

**Reply by pr-ju** [reply]
My idea was in order to keep the effort minimal using the official pre-existing container images of debian-slim that were provided by for example in the docker registry.  I pulled this via nerdctl and saved the container image to the host machine (M1 Pro CPU) via nerdctl save 

`nctl save debian:bullseye-slim -o debian_slim.img --platform=arm64`

and then tried it to boot. 

`limactl start --name debian_slim  debian_slim.yaml`

But It did not do it (see serial.log below). I guess the img is a tar actually and the format is not right somehow. 

debian_slim.yaml:
------------------------
```
vmType: "qemu"
arch: "aarch64"
images:
- location: ""~/debian_slim.img""
  arch: "aarch64"

memory: "2GiB"

containerd:
  system: false
  user: false

mounts:
- location: "~"
  writable: true
```


serial.log
------------------------
```
UEFI firmware (version edk2-stable202302-for-qemu built at 17:14:55 on Mar 15 2023)
Error: Image at 000BFC89000 start failed: Unsupported
Error: Image at 000BFC11000 start failed: Not Found
Tpm2SubmitCommand - Tcg2 - Not Found
Tpm2GetCapabilityPcrs fail!
Tpm2SubmitCommand - Tcg2 - Not Found
Image type X64 can't be loaded on AARCH64 UEFI system.
[2J[01;01H[=3h[2J[01;01H[2J[01;01H[=3h[2J[01;01HBdsDxe: failed to load Boot0001 "UEFI QEMU QEMU CD-ROM " from PciRoot(0x0)/Pci(0x1,0x0)/Scsi(0x0,0x0): Not Found
BdsDxe: failed to load Boot0002 "UEFI Misc Device" from PciRoot(0x0)/Pci(0x5,0x0): Not Found

>>Start PXE over IPv4.
```

**Reply by afbjorklund** [reply]
You seem to be mixing container images (.tar) with virtual machine images (.img), those are totally different concepts...
Somewhat simplified you could say that we take a rootfs (.tar), add a kernel to it, make it bootable and create an image.

```
REPOSITORY   TAG                 IMAGE ID       CREATED         SIZE
debian       stable              051447af7d04   4 days ago      124MB
debian       stable-slim         cc2f487c8222   4 days ago      80.5MB
```

```
#	/home/anders/.cache/lima/download/by-url-sha256/37f5fc187ab151950c3e7f30e2231bedf2f8af23b24933aec9f3a2a2020fa4ba
319M	https://cloud.debian.org/images/cloud/bullseye/20230124-1270/debian-11-generic-amd64-20230124-1270.qcow2
```

We also have CD/DVD images (.iso), that we use for live booting and for cloud-init configuration, another different format.
They work somewhat similar to the HD images, except that they are read-only. That's why we have "basedisk" and "diffdisk"

**Reply by afbjorklund** [reply]
But you can start from `template://debian`, and experiment with your improvements from there ?

Currently there is no good way to "capture" the result, but I guess can save copies of the disk images.

A reasonable expection of the disk image size (from before), was 100M for boot2docker (compared to 50M)...

That is, the debian-based image (20.10)  was 100M while the tinycore-based image (19.03) was around 50M

**Reply by pr-ju** [reply]
Concerning debian images sizes: On https://cloud.debian.org/images/cloud/bullseye/latest/ all qcow2 images are at least 300 MiB, so smaller, you are right, of course, but still quite large compared to alpine. The odd thing here is that Debian-slim is not available at the official Debian servers (or I could not find it), only in the container repositories. 

The repository for providing the "official" Dockerfile is here: 

https://github.com/debuerreotype/docker-debian-artifacts/blob/4648ad5c6feb1318b8a6c31502ebde8bb18ebb5e/bullseye/slim/Dockerfile

Even the rootfs.tar.xz that they use in the Dockerfile is there. But since I'm newbie here coming from vagrant with Vbox Backend I do not know how to convert it to use it in lima.

Of course, the debian-slim container images work well inside nerdctl in the default vm And the sizes are also good. But that is actually not what I want. I only need 4 to 5 five VMs for developments projects with a small size, which could altered from time to time by directly accessing them via shell. For that purpose lima is really perfect(kudos to you for this project), since  I do not need a fully flavored container system, which add another level of complexity I do not need.    

```
REPOSITORY    TAG                   IMAGE ID        CREATED         PLATFORM          SIZE         BLOB SIZE
debian        bullseye-slim         7606bef5684b    19 hours ago    linux/arm64/v8    81.8 MiB     28.7 MiB
node          bullseye-slim         dc1906714d19    14 hours ago    linux/arm64/v8    259.4 MiB    77.1 MiB
python        3.11-slim-bullseye    eaee5f73efa9    13 hours ago    linux/arm64/v8    132.7 MiB    44.4 MiB
```

**Reply by afbjorklund** [reply]
As far as I know, the "slim" is a term from Docker regarding the package selection: https://hub.docker.com/_/debian

> These tags are an experiment in providing a slimmer base (removing some extra files that are normally not necessary within containers, such as man pages and documentation), and are definitely subject to change.

> See the debuerreotype-slimify script (debuerreotype linked above) for more details about what gets removed during the "slimification" process.

You can use something like `container-diff`, to compare the end result between the two images

EDIT: This is the script: https://github.com/debuerreotype/debuerreotype/blob/master/scripts/debuerreotype-slimify

* https://github.com/debuerreotype/debuerreotype/blob/master/scripts/.slimify-includes
* https://github.com/debuerreotype/debuerreotype/blob/master/scripts/.slimify-excludes

**Reply by pr-ju** [reply]
> But you can start from `template://debian`, and experiment with your improvements from there ?
> 
> Currently there is no good way to "capture" the result, but I guess can save copies of the disk images.
> 
> A reasonable expection of the disk image size (from before), was 100M for boot2docker (compared to 50M)...
> 
> That is, the debian-based image (20.10) was 100M while the tinycore-based image (19.03) was around 50M

You were faster than my other comment  about the image sizes. I will try the images you mentioned, I was not able to find them previously, sorry. Do you have the links?

**Reply by afbjorklund** [reply]
You mean for https://github.com/boot2docker/boot2docker and https://github.com/tianon/boot2docker-debian ?

They were mostly for reference/comparison; they have not been updated in years, are x86 only, and only have docker

**Reply by afbjorklund** [reply]
My thinking was that the lima-alpine was "good enough", to haven't worked so much on the alternative OS.

Especially if you require using rootless containers, since then you are forced to use systemd (and also glibc)

**Reply by pr-ju** [reply]
> https://github.com/debuerreotype/docker-debian-artifacts/blob/4648ad5c6feb1318b8a6c31502ebde8bb18ebb5e/bullseye/slim/Dockerfile

No I was just referring to the locations of the two images you mentioned:

> debian-based image (20.10) was 100M while the tinycore-based image (19.03) was around 50M

as well to the repository where the debian-slim containers are administered for the docker registry

https://github.com/debuerreotype/docker-debian-artifacts/blob/4648ad5c6feb1318b8a6c31502ebde8bb18ebb5e/bullseye/slim/

**Reply by afbjorklund** [reply]
I don't have the old images up anymore, sorry. They were just ballpark figures on what I remembered by heart.

The "docker-machine" and "boot2docker" projects are only related in spirit to Lima, the code is completely different.

**Reply by afbjorklund** [reply]
I would suggest creating a github issue, regarding the possible creation of such a debian-based image for Lima ?

Most likely, it would be a separate project (similar to https://github.com/lima-vm/alpine-lima) and need volunteers.

**Reply by pr-ju** [reply]
> I would suggest creating a github issue, regarding the possible creation of such a debian-based image for Lima ?

You mean here in the project?

> Most likely, it would be a separate project (similar to https://github.com/lima-vm/alpine-lima) and need volunteers.

I would even trying to support, but my current skill-set concerning the topic is not very well developed, but I'm trying to make myself familiar with it.

**Reply by afbjorklund** [reply]
There are several different aspects of this particular topic...

The first would be to change the base of your container images, to use "debian-slim" instead of "alpine" as per those articles linked. Related to that is some documentation in Lima, regarding what is the diffence between using Alpine VM and using Debian(/Ubuntu) VM. Such a documentation would have to explain musl libc vs. glibc, and openrc vs. systemd. Something like a pros and cons? Most of this can be accomplished without changing any code in the Lima project itself, and many of the topics are more generic in nature.

Then it comes to the issue if the Lima project should provide a custom version of Debian(/Ubuntu), like it provides a custom version of Alpine? For alpine-lima, it is necessary because certain features are missing from Alpine "virt". But for debian-lima, it would be more of an optimization (like that "slimming" script) since all required features are already in Debian(/Ubuntu) cloud images ("NoSource"). So it is not a given, that it is worth the extra work of maintaining a downstream distribution - just to save disk space.

So it would help to formulate, what changes you were thinking of.
